### PR TITLE
feat: add class named & string/class qualifier

### PR DESCRIPTION
**Pull request by ghasemdev**

### CODE REVIEW

**Comment by MaxBQb** [comment]
Can you also support [complex types](https://github.com/InsertKoinIO/koin-annotations/issues/77#issuecomment-1868312179)?

**Comment by ghasemdev** [comment]
> Can you also support [complex types](https://github.com/InsertKoinIO/koin-annotations/issues/77#issuecomment-1868312179)?

Why do you need such a complex structure?

**Comment by MaxBQb** [comment]
Well, imagine `Mapper<MyDomain, MyView>`
we inject it using `@Factory` and we also need to add type qualifier, and just `MyDomain` alone isn't enough.
Or Imagine we use `Serializer<List<MyDomain>>`, so we need to add type qualifier not for `List`, not for `MyDomain`, but for `List<MyDomain>`, so here tree-structure is preferable for such usecases (and not flat one, because we must distinguish `SomeClass<Int, OtherClass<Int>>` and `SomeClass<Int, OtherClass, Int>`)
Such tree structure required only at annotation definition, and then can be converted to string for example, to be easily compared (see code I linked above).

So that's why I think that if we adding support for type-based qualifiers, then we need to at least support generic-types.

**Comment by ghasemdev** [comment]
> Well, imagine `Mapper<MyDomain, MyView>` we inject it using `@Factory` and we also need to add type qualifier, and just `MyDomain` alone isn't enough. Or Imagine we use `Serializer<List<MyDomain>>`, so we need to add type qualifier not for `List`, not for `MyDomain`, but for `List<MyDomain>`, so here tree-structure is preferable for such usecases (and not flat one, because we must distinguish `SomeClass<Int, OtherClass<Int>>` and `SomeClass<Int, OtherClass, Int>`) Such tree structure required only at annotation definition, and then can be converted to string for example, to be easily compared (see code I linked above).
> 
> So that's why I think that if we adding support for type-based qualifiers, then we need to at least support generic-types.

### possible solution:
#### Nested Structure
```kotlin
sealed interface ListQualifier {
  data object Int : ListQualifier
  sealed interface Str : ListQualifier {
    data object String : Str
  }
}

@Module
class ListModule {
  @Singleton(createdAtStart = true)
  @Qualifier(ListQualifier.Int::class)
  fun provideIntList() = listOf(1, 2, 3)

  @Singleton(createdAtStart = true)
  @Qualifier(ListQualifier.Str.String::class)
  fun provideStringList() = listOf("hi", "hello")
}

class MyActivity : ActivityScope() {
  private val intList: List<Int> by inject(qualifier<ListQualifier.Int>())
  private val stringList: List<String> by inject(qualifier<ListQualifier.Str.String>())
}
```
#### Flat Structure
```kotlin
annotation class IntListQualifier
annotation class StringListQualifier

@Module
class ListModule {
  @Singleton(createdAtStart = true)
  @Qualifier(IntListQualifier::class)
  fun provideIntList() = listOf(1, 2, 3)

  @Singleton(createdAtStart = true)
  @Qualifier(StringListQualifier::class)
  fun provideStringList() = listOf("hi", "hello")
}

class MyActivity : ActivityScope() {
  private val intList: List<Int> by inject(qualifier<IntListQualifier>())
  private val stringList: List<String> by inject(qualifier<StringListQualifier>())
}

### Generated Code
#### Nested Structure
```kotlin
single(qualifier=org.koin.core.qualifier.StringQualifier("com.parsuomash.sdk.di.qualifier.ListQualifier\$Int"),createdAtStart=true) { moduleInstance.provideIntList() } bind(kotlin.collections.List::class)
	single(qualifier=org.koin.core.qualifier.StringQualifier("com.parsuomash.sdk.di.qualifier.ListQualifier\$Str\$String"),createdAtStart=true) { moduleInstance.provideStringList() } bind(kotlin.collections.List::class)
}
```
#### Flat Structure
```kotlin

```

```

**Comment by MaxBQb** [comment]
Well that looks kinda hacky. The flat structure looks interesting (except it quickly gets verbose).
But it does not create clear connection between real kotlin types and qualifier, so it will be hard to query instances later on.
I mean, while simple querying is still trivial, it will be quite hard to retrieve `Mapper<Domain, View>` for specific `Domain` and `View`, known at compile-time. You'll end up building yet another big `when` condition to get `MapperDomain1View1`, ``MapperDomain2View2`` ... or `Mapper.Domain1.View1`, `Mapper.Domain2.View2`.
With my solution, on the other hand you'll still be able to use `FlatMeaninglessAnnotation` for trivial usecases, but you will also able to (with the help of an inline function) retrieve instances for class with specific generics. So I can write one little function  `getMapper` which will build `TypeQualifier` and quickly find what I need.

inline function (aka `typeQualifierOf`) should be part of API (I'll try to write/suggest one, later on).

**Comment by MaxBQb** [comment]
Also please provide any feedback on my reasoning and code above, because, well, I see reflection of our discussion in commit name, but not in code changes itself.

**Comment by ghasemdev** [comment]
> Also please provide any feedback on my reasoning and code above, because, well, I see reflection of our discussion in commit name, but not in code changes itself.

I think this method can also be implemented in another PR.

@arnaudgiuliani

**Comment by MaxBQb** [comment]
Anyway your current annotation can't be used to build complex type structure (yeah that recursive-annotation thing looks even more horrible than my DSL), am I right?

**Comment by MaxBQb** [comment]
I should clarify my intentions: even if u don't want to implement support for complex types by yourself, please at least add support for that at annotation side, so next changes can be implemented incrementally, without need of breaking existing interface (without need of rewrite TypeQualifier annotation)

**Comment by MaxBQb** [comment]
Ghm.. [this whole time](https://github.com/InsertKoinIO/koin-annotations/pull/108#issuecomment-1869408254) we can just use `typeOf`... My bad, I've almost forgot about such simple thing.

```kt
package org.example

import kotlin.reflect.KClass
import kotlin.reflect.KType
import kotlin.reflect.jvm.jvmErasure
import kotlin.reflect.typeOf

annotation class TypeQualifier(val value: KClass<*>, vararg val params: TypeQualifier)

val TypeQualifier.qualifier: String get() {
    var qualifier = value.qualifiedName!!
    if (params.isNotEmpty())
        qualifier += "<" + params.joinToString { it.qualifier } + ">"
    return qualifier
}

fun typeQualifierOf(type: KType): TypeQualifier
    = TypeQualifier(type.jvmErasure, *type.arguments.map {
        typeQualifierOf(it.type!!)
    }.toTypedArray())
inline fun <reified T : Any> typeQualifierOf()= typeQualifierOf(typeOf<T>())

fun main() {
    println(typeQualifierOf<Map<Map<Int, String>, String>>().qualifier)
}
```

**Comment by MaxBQb** [comment]
So.. now such simple thing can be implemented in the scope of the same PR, what do you think?

**Comment by Fando** [comment]
Would be valuable if defining type annotations could be less verbose. Usually a qualifying annotation is used in many places in the code causing it to look cluttered.

Instead of this:
```
@Scope(name = "scope1")  

@Named(name = "name1") 
```

Prefer less verbose definitions like:
```
annotation class Scope1

annotation class Name1
```
This way `@Scope1` could be used instead of `@Scope(name = "scope1")`

File: compiler/koin-ksp-compiler/src/jvmMain/kotlin/org/koin/compiler/verify/KoinConfigVerification.kt [hunk 1]
```diff
@@ -24,7 +24,12 @@ import org.koin.compiler.generator.getFile
 import org.koin.compiler.metadata.KoinMetaData
 import java.io.OutputStream
 
-private val ignored = listOf("kotlin.Any",)
+private val ignored = listOf(
+    "kotlin.Any",
+    "androidx.lifecycle.SavedStateHandle",
+    "android.content.Context",
+    "android.app.Application"
+)
```
Comment (arnaudgiuliani) [inline compiler/koin-ksp-compiler/src/jvmMain/kotlin/org/koin/compiler/verify/KoinConfigVerification.kt:32]: seems out of scope?

**Review by arnaudgiuliani — APPROVED** [review]
Good job

**Comment by arnaudgiuliani** [comment]
same about preparing 1.4.0 branch - https://github.com/InsertKoinIO/koin-annotations/pull/107

**Comment by Neyasbit** [comment]
I tried to do that. And it only works 

```kotlin
@Named("CoroutineDispatcherIO")
annotation class NamedCoroutineDispatcherIO

@Named("CoroutineDispatcherMain")
annotation class NamedCoroutineDispatcherMain
@Single
@NamedCoroutineDispatcherIO
fun provideDispatcherIO(): CoroutineDispatcher = Dispatchers.IO

@Single
@NamedCoroutineDispatcherMain
fun provideDispatcherMain(): CoroutineDispatcher = Dispatchers.Main
```
But there is a bug. When using two annotations at the same time. It turns out that there will be an instance of one object. Although it should be different. 

```kotlin
class Test(
    @NamedCoroutineDispatcherIO dispatcherIO: CoroutineDispatcher,
    @NamedCoroutineDispatcherMain dispatcherMain: CoroutineDispatcher
) {
    init {
        // Wrong
        println(dispatcherIO.hashCode() == dispatcherMain.hashCode()) // True (expect false)
    }
}
```

**Comment by extmkv** [comment]
@ghasemdev when are you planning to merge it?
